src/librustc_lint/types.rs

```diff
@@ -695,11 +705,20 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
                         // Check the contained variants.
                         for variant in &def.variants {
                             for field in &variant.fields {
-                                let arg = cx.normalize_erasing_regions(
+                                let field_ty = cx.normalize_erasing_regions(
                                     ParamEnv::reveal_all(),
                                     field.ty(cx, substs),
                                 );
-                                let r = self.check_type_for_ffi(cache, arg);
+                                if def.repr.transparent() {
+                                    let is_zst = cx
```

Comment:
It'd be helpful to include the corresponding comment about why ZST fields can be skipped.